### PR TITLE
Fix #4417: Do not track `Error` as dangerous global ref.

### LINKED ISSUE #4417: The emitter systematically goes into the secondAttempt on the first run
It tracks the global ref `Error`, which is used as super class of `Throwable`, and does not filter it out as non-dangerous.

### PR DESCRIPTION
**Pull request by sjrd**

### CODE REVIEW

File: linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/JSGen.scala [hunk 1]
```diff
@@ -133,6 +133,10 @@ private[emitter] final class JSGen(val config: Emitter.Config) {
   def globalRef(name: String)(implicit pos: Position): WithGlobals[VarRef] =
     WithGlobals(VarRef(Ident(name)), Set(name))
 
+  def untrackedGlobalRef(name: String)(implicit pos: Position): WithGlobals[VarRef] =
```
Comment (gzm0) [inline linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/JSGen.scala:136]: Assert that the global ref is not dangerous?

Also, de-duplicate with the code in `CoreJSLib`? (I realize this means more allocations if `WithGlobals` cannot be stack allocated, but I don't think that matters in the `CoreJSLib`).

File: linker/shared/src/test/scala/org/scalajs/linker/LinkerTest.scala [hunk 2]
```diff
@@ -50,6 +52,39 @@ class LinkerTest {
...
+    val PrintStreamClass = ClassName("java.io.PrintStream")
+    val PrintStreamType = ClassType(PrintStreamClass)
+
+    val classDefs = List(
+      mainTestClassDef {
+        val err = ApplyStatic(EAF, "java.lang.System",
+            m("err", Nil, ClassRef(PrintStreamClass)), Nil)(PrintStreamType)
+        Apply(EAF, err, m("println", List(O), V), List(str("Hello world!")))(NoType)
+      }
+    )
```
Comment (gzm0) [inline linker/shared/src/test/scala/org/scalajs/linker/LinkerTest.scala:70]: Just use this instead? https://github.com/scala-js/scala-js/blob/92c64df83377d47c55d3de0ff78d10da3ba06bcc/linker/shared/src/test/scala/org/scalajs/linker/OptimizerTest.scala#L179-L183

File: linker/shared/src/test/scala/org/scalajs/linker/LinkerTest.scala [hunk 3]
```diff
@@ -50,6 +52,39 @@ class LinkerTest {
...
+    val classDefsFiles = classDefs.map(MemClassDefIRFile(_))
+
+    for {
+      fulllib <- TestIRRepo.fulllib
+      report <- linker.link(fulllib ++ classDefsFiles,
+          MainTestModuleInitializers, MemOutputDirectory(), logger)
+    } yield {
+      val log = logger.allLogLines
+      assertFalse("The second attempt of the emitter should not be triggered\n" + log,
+          log.contains("Emitter: The set of dangerous global refs has changed."))
```
Comment (gzm0) [inline linker/shared/src/test/scala/org/scalajs/linker/LinkerTest.scala:84]: Consider adding a test that tests this error message is present if a dangerous global ref is present. That way, we avoid the test becoming useless (without just relying on the comment).

File: linker/shared/src/test/scala/org/scalajs/linker/LinkerTest.scala [hunk 4]
```diff
@@ -50,6 +52,39 @@ class LinkerTest {
     testLink(helloWorldClassDefs, MainTestModuleInitializers)
   }
 
+  /** Makes sure that linking a "substantial" program (using `println`) does
+   *  not trigger the second attempt in the Emitter due to dangerous global
+   *  refs.
+   */
+  @Test
+  def linkNoSecondAttemptInEmitter(): AsyncResult = await {
```
Comment (gzm0) [inline linker/shared/src/test/scala/org/scalajs/linker/LinkerTest.scala:60]: Put this in a new class `EmitterTest`? Even though it goes through the interface of the linker, it tests the emitter (we have a lot of precedent in tests here to do that).

File: linker/shared/src/test/scala/org/scalajs/linker/testutils/CapturingLogger.scala [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+ * additional information regarding copyright ownership.
+ */
+
+package org.scalajs.linker.testutils
+
+import scala.collection.mutable
+
+import org.scalajs.logging._
+
+final class CapturingLogger extends Logger {
```
Comment (gzm0) [inline linker/shared/src/test/scala/org/scalajs/linker/testutils/CapturingLogger.scala:19]: Deduplicate this with the ad-hoc implementation in `IRCheckerTest`?

File: linker/shared/src/test/scala/org/scalajs/linker/testutils/TestIRBuilder.scala [hunk 6]
```diff
@@ -134,4 +134,5 @@ object TestIRBuilder {
     MethodIdent(name)
 
   def int(x: Int): IntLiteral = IntLiteral(x)
+  def str(x: String): StringLiteral = StringLiteral(x)
```
Comment (gzm0) [inline linker/shared/src/test/scala/org/scalajs/linker/testutils/TestIRBuilder.scala:137]: If you introduce this, "backfill" usage in another commit? (I can find 9 occurrences of `StringLiteral` in linker tests).

**Review by gzm0 — CHANGES_REQUESTED** [review]

File: linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/JSGen.scala (hunk 1 above)
Comment (sjrd) [reply linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/JSGen.scala]: > Assert that the global ref is not dangerous?

Good point.

> Also, de-duplicate with the code in `CoreJSLib`? (I realize this means more allocations if `WithGlobals` cannot be stack allocated, but I don't think that matters in the `CoreJSLib`).

I tried that, but in `CoreJSLib` the assertion is stronger than what this achieves: it asserts that *all* global refs emitted by the `CoreJSLib` (including through any method that returns `WithGlobal`s, not only `globalRef`) are non-dangerous. Using this new method would be weaker.

**Review by sjrd — COMMENTED** [review]
Updated.

**Review by gzm0 — APPROVED** [review]

**Pull request merged**